### DISCUSSION TITLE
Enable custom registry support for the `node-version` datasource

### DISCUSSION THREAD

**Discussion by bienzaaron** [renovatebot/renovate#25123] [Suggest an Idea]
Related discussion https://github.com/renovatebot/renovate/discussions/17518

My team uses the https://github.com/containerbase/base docker image as a part of our self-hosted Mend + Renovate. We are running into artifact update issues on renovate PRs for npm/yarn when [`binarySource`](https://docs.renovatebot.com/self-hosted-configuration/#binarysource) is set to `install` (the default), because the node-version datasource doesn't support custom registries, and accessing https://nodejs.org is blocked by our corporate proxy.

Would it make sense to flip the `customRegistrySupport` flag [here](https://github.com/renovatebot/renovate/blob/f31c0883c737c7e7700762cc35e2e5b3f7e29c23/lib/modules/datasource/node-version/index.ts#L16) to true? At least for us, it would solve our problem, because we internally mirror `https://nodejs.org/dist` and could use that instead. I imagine other enterprise users may have a similar need.

As far as I can tell, flipping this flag + updating docs would be the only change required (everything else looks like it should just work, but I'm not super familiar with the codebase). Happy to open a PR if that's the case.

**Comment by rarkins** [comment]
We have a lot of datasources with customRegistrySupport=false:

<img width="320" alt="image" src="https://github.com/renovatebot/renovate/assets/6311784/32ff1f53-68c5-4338-aa82-04e1a965542d">

The idea of this was to prevent users from shooting themselves in the foot with invalid settings. In your case if you block all URLs by default then you might need the ability to rewrite _most_ of these at some point in the future, so I'm not sure whether to say "sure, change this one" or whether we need a more holistic approach.

**Comment by bienzaaron** [comment]
I set up a [test repo](https://github.com/bienzaaron/renovate-test-repo/blob/main/renovate.json#L14) and blocked nodejs.org locally, and it turns out just flipping the flag doesn't work for the dynamic install bit anyway.

It works fine for actually updating the node version (using asdf, since that uses the `node-version` datasource), but not installing node as a pre-step to running npm/yarn - I'm guessing `packageRules` doesn't get applied in that case (which makes sense, I think).

**relevant log bits**

```
DEBUG: Using containerbase dynamic installs (repository=bienzaaron/renovate-test-repo, baseBranch=main, branch=renovate/node-18.x)
DEBUG: GET https://nodejs.org/dist/index.json = (code=ECONNREFUSED, statusCode=-1 retryCount=2, duration=2) (repository=bienzaaron/renovate-test-repo, baseBranch=main, branch=renovate/node-18.x)
DEBUG: Datasource connection error (repository=bienzaaron/renovate-test-repo, baseBranch=main, branch=renovate/node-18.x)
       "datasource": "node-version",
       "packageName": "node",
       "url": undefined,
       "errCode": "ECONNREFUSED"
 WARN: No tool releases found. (repository=bienzaaron/renovate-test-repo, baseBranch=main, branch=renovate/node-18.x)
       "toolConfig": {"datasource": "node-version", "packageName": "node", "versioning": "node"}
DEBUG: lock file error (repository=bienzaaron/renovate-test-repo, baseBranch=main, branch=renovate/node-18.x)
       "err": {
         "message": "No tool releases found.",
         "stack": "Error: No tool releases found.\n    at resolveConstraint (/home/ubuntu/renovate/lib/util/exec/containerbase.ts:273:11)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at async generateInstallCommands (/home/ubuntu/renovate/lib/util/exec/containerbase.ts:324:27)\n    at async prepareRawExec (/home/ubuntu/renovate/lib/util/exec/index.ts:120:11)\n    at async exec (/home/ubuntu/renovate/lib/util/exec/index.ts:146:39)\n    at async Object.generateLockFile (/home/ubuntu/renovate/lib/modules/manager/npm/post-update/npm.ts:153:5)\n    at async getAdditionalFiles (/home/ubuntu/renovate/lib/modules/manager/npm/post-update/index.ts:506:17)\n    at async processBranch (/home/ubuntu/renovate/lib/workers/repository/update/branch/index.ts:471:31)\n    at async writeUpdates (/home/ubuntu/renovate/lib/workers/repository/process/write.ts:159:17)\n    at async update (/home/ubuntu/renovate/lib/workers/repository/process/extract-update.ts:214:11)\n    at async Object.renovateRepository (/home/ubuntu/renovate/lib/workers/repository/index.ts:76:19)\n    at async attributes.repository (/home/ubuntu/renovate/lib/workers/global/index.ts:184:11)\n    at async start (/home/ubuntu/renovate/lib/workers/global/index.ts:169:7)\n    at async /home/ubuntu/renovate/lib/renovate.ts:18:22"
       },
       "type": "npm"
```

http stats (shows hitting both nodejs.org for the dynamic install bit and my configured mirror - `mynodemirror.com` for actually updating the node version):
```
DEBUG: http statistics (repository=bienzaaron/renovate-test-repo)
       "urls": {
         "http://mynodemirror.com:3000/some/dir/index.json (GET,200)": 1,
         "http://mynpmregistry.com:3000/some/other/dir/fastify (GET,200)": 1,
         ... bunch of github stats....,
         "https://nodejs.org/dist/index.json (GET,-1)": 1
       },
       "hostStats": {
         "mynodemirror.com": {"requestCount": 1, "requestAvgMs": 91, "queueAvgMs": 0},
         "mynpmregistry.com": {
           "requestCount": 1,
           "requestAvgMs": 172,
           "queueAvgMs": 0
         },
         "api.github.com": {"requestCount": 40, "requestAvgMs": 739, "queueAvgMs": 0},
         "nodejs.org": {"requestCount": 1, "requestAvgMs": 2, "queueAvgMs": 0}
       },
       "totalRequests": 43
```

**Reply by rarkins** [reply]
I think you're better for now to use the "-full" Docker image and set RENOVATE_BINARY_SOURCE=global so that dynamic versions aren't installed.

Related: https://github.com/containerbase/base/blob/main/docs/custom-registries.md

**Reply by bienzaaron** [reply]
My only reservation would be if some repo requires a particular node major. To get around that, I think we could:
1. install the latest of each major in our custom image
2. set `binarySource` to `global`
then do something like:
```
{
  "packageRules": [
    {
      "matchDatasource": ["npm", "yarn"],
      "postUpgradeTasks": {
        "commands": [
          "NODE_MAJOR=16 install-tool node v$(ls /opt/buildpack/tools/node | grep ^$NODE_MAJOR | sort -V)",
        ]
      }
    }
  ]
}
```

as long as we're using a version of node already installed in the image, should work fine. I _think_ that should get us what we need for now. I appreciate your help 😄

**Reply by bienzaaron** [reply]
Ah, nvm. think those get run after npm/yarn locks are updated.
